### PR TITLE
[SPARK-38583][SQL] Restore the behavior of `to_timestamp` that allows numeric types

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR is a followup of https://github.com/apache/spark/commit/fab4ceb157baac870f6d50b942084bb9b2cd4ad2 that mistakenly removed the numeric type support in `to_timestamp(...)`. This PR restores the behavior back.

### Why are the changes needed?

To keep the previous behavior.

### Does this PR introduce _any_ user-facing change?

To end users, no because the change has not been released yet.

### How was this patch tested?

Unit test was added, and manually tested:

```scala
spark.range(1).selectExpr("to_timestamp(id)").show()
```

**Before**

```
+----------------+
|to_timestamp(id)|
+----------------+
|            null|
+----------------+
```

**After**

```
+-------------------+
|   to_timestamp(id)|
+-------------------+
|1970-01-01 09:00:00|
+-------------------+

```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @cloud-fan and @xinrong-databricks FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -2080,8 +2080,10 @@ case class ParseToTimestamp(
   override def inputTypes: Seq[AbstractDataType] = {
     // Note: ideally this function should only take string input, but we allow more types here to
     // be backward compatible.
-    TypeCollection(StringType, DateType, TimestampType, TimestampNTZType) +:
-      format.map(_ => StringType).toSeq
+    val types = Seq(StringType, DateType, TimestampType, TimestampNTZType)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2083]: BTW, it's from `Cast` logic - we used `Cast` here before https://github.com/apache/spark/commit/fab4ceb157baac870f6d50b942084bb9b2cd4ad2:
https://github.com/apache/spark/blob/c1e5e8a27595c711c53bb3571ea391ef700d531a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L71

**Review by cloud-fan — APPROVED** [review]
good catch!

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.3.

**Comment by xinrong-databricks** [comment]
Thanks @HyukjinKwon !

**Pull request closed without merging**